### DISCUSSION TITLE
Dynamic Enums

### DISCUSSION THREAD

**Discussion by thomascobb** [wyfo/apischema#128] [Q&A]
I would like to make a dynamic enum, where the values are based on the presence of keys in a document, something like https://github.com/json-schema-org/json-schema-spec/issues/706.

I have read through the associated issues on json-schema-org, and they all seem to leave it up to the implementation as to whether/how this should be implemented:

> While none of the vocabularies defined as part of this or the associated documents
> define a keyword which may target and/or load instance data, it is possible that
> other vocabularies may wish to do so.
https://github.com/json-schema-org/json-schema-spec/pull/1034/files

I could do this outside apischema by post-processing the deserialized objects, but I wondered if this was within the scope of what apischema could do, and if you'd given it any thought?

Thanks,
Tom

**Comment by wyfo** [comment]
You can use the deserialized data during validation using [validators](https://wyfo.github.io/apischema/validation/#dataclass-validators).
Here is a simple example:
```python
from dataclasses import dataclass, field
from apischema import ValidationError, deserialize, validator, serialize


@dataclass
class Data:
    allowed: list[str]
    values: list[str] = field()

    @validator(values)
    def check_values_are_allowed(self):
        for index, value in enumerate(self.values):
            if value not in self.allowed:
                yield index, f"value not allowed, should be one of {self.allowed}"


assert deserialize(
    Data, {"allowed": ["foo", "bar"], "values": ["foo", "bar", "foo"]}
) == Data(["foo", "bar"], ["foo", "bar", "foo"])
try:
    deserialize(Data, {"allowed": ["foo"], "values": ["foo", "bar", "foo"]})
except ValidationError as err:
    assert serialize(err) == [
        {"loc": ["values", 1], "err": ["value not allowed, should be one of ['foo']"]}
    ]
```

Moreover, when it comes to schema generation, it's possible to add a `$data` keyword  using [schema extra](https://wyfo.github.io/apischema/json_schema/#schema-annotations) — it will be easier in version 0.15 using callable extra.

**Reply by thomascobb** [reply]
Thanks for this, it's good to know this can be accomplished with validators. I'll be interested to see how the schema with `$data` looks. Do you know of any editors that use `$data` yet? I use https://github.com/redhat-developer/vscode-yaml for editing YAML files backed with a JSON schema, but I can't see any mention of `$data` in their codebase so I guess it doesn't yet.

**Reply by wyfo** [reply]
> Do you know of any editors that use `$data` yet?

Unfortunately, no. I only know [AJV](https://github.com/ajv-validator/ajv), but it's a validator. By the way, because the `$data` keyword proposal was rejected, I don't think it will gain a lot of support.

> I'll be interested to see how the schema with `$data` looks.

Something like that — it requires 0.15 (which, I hope, will be released soon):
```python
from dataclasses import dataclass, field
from apischema import ValidationError, deserialize, validator, schema, serialize
from apischema.json_schema import deserialization_schema


@schema(
    extra=lambda s: s["properties"]["values"].update({"enum": {"$data": "/allowed"}})
)
@dataclass
class Data:
    allowed: list[str]
    values: list[str] = field()

    @validator(values)
    def check_values_are_allowed(self):
        for index, value in enumerate(self.values):
            if value not in self.allowed:
                yield index, f"value not allowed, should be one of {self.allowed}"


assert deserialization_schema(Data) == {
    "$schema": "http://json-schema.org/draft/2019-09/schema#",
    "type": "object",
    "properties": {
        "allowed": {"type": "array", "items": {"type": "string"}},
        "values": {
            "type": "array",
            "items": {"type": "string"},
            "enum": {"$data": "/allowed"},
        },
    },
    "required": ["allowed", "values"],
    "additionalProperties": False,
}
```
